In [54]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import run_tests_params
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc read-aloud"><h2>--- Day 20: Firewall Rules ---</h2><p>You'd like to set up a small hidden computer here so you can use it to <span title="I'll create a GUI interface using Visual Basic... see if I can track an IP address.">get back into the network</span> later. However, the corporate firewall only allows communication with certain external <a href="https://en.wikipedia.org/wiki/IPv4#Addressing">IP addresses</a>.</p>
<p>You've retrieved the list of blocked IPs from the firewall, but the list seems to be messy and poorly maintained, and it's not clear which IPs are allowed. Also, rather than being written in <a href="https://en.wikipedia.org/wiki/Dot-decimal_notation">dot-decimal</a> notation, they are written as plain <a href="https://en.wikipedia.org/wiki/32-bit">32-bit integers</a>, which can have any value from <code>0</code> through <code>4294967295</code>, inclusive.</p>
<p>For example, suppose only the values <code>0</code> through <code>9</code> were valid, and that you retrieved the following blacklist:</p>
<pre><code>5-8
0-2
4-7
</code></pre>
<p>The blacklist specifies ranges of IPs (inclusive of both the start and end value) that are <em>not</em> allowed. Then, the only IPs that this firewall allows are <code>3</code> and <code>9</code>, since those are the only numbers not in any range.</p>
<p>Given the list of blocked IPs you retrieved from the firewall (your puzzle input), <em>what is the lowest-valued IP</em> that is not blocked?</p>
</article>


In [55]:
def lowest_non_blocked_ip(blacklist: str) -> int:
    blocked = []
    for line in blacklist.strip().splitlines():
        fr, to = map(int, line.split("-"))
        blocked.append((fr, to + 1))

    blocked.sort()
    to_prev = 0
    for fr, to in blocked:
        if fr - to_prev > 0:
            return to_prev
        to_prev = to
    return -1


blacklist_example = """
5-8
0-2
4-7
"""


lowest_non_blocked_ip(blacklist_example), 3

(3, 3)

In [56]:
with open("../input/day20.txt") as f:
    blacklist_partI = f.read()

lowest_non_blocked_ip(blacklist_partI)

19449262

<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>19449262</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p><em>How many IPs</em> are allowed by the blacklist?</p>
</article>

</main>


In [57]:
def count_allowed(blacklist: str, fr: int, to: int) -> int:
    blocked = []
    for line in blacklist.strip().splitlines():
        fr, to = map(int, line.split("-"))
        blocked.append((fr, to + 1))

    blocked.sort()
    allowed = 0
    to_prev = 0

    for fr, to in blocked:
        if fr - to_prev > 0:
            allowed += fr - to_prev
        to_prev = max(to_prev, to)

    if to_prev < to:
        allowed += to - to_prev
    return allowed


count_allowed(blacklist_example, 0, 9)

1

In [58]:
count_allowed(blacklist_partI, 0, 19449262)

119

<link href="style.css" rel="stylesheet"></link>

<p>Your puzzle answer was <code>119</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>
